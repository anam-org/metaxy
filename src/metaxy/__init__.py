from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
