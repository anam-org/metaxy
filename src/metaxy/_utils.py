import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
