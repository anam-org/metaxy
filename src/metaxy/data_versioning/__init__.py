"""Data versioning module for sample-level data version calculation."""

from metaxy.data_versioning.calculators import (
    DataVersionCalculator,
    PolarsDataVersionCalculator,
)
from metaxy.data_versioning.diff import (
    DiffResult,
    MetadataDiffResolver,
    PolarsDiffResolver,
)
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.data_versioning.joiners import PolarsJoiner, UpstreamJoiner

__all__ = [
    "HashAlgorithm",
    "UpstreamJoiner",
    "PolarsJoiner",
    "DataVersionCalculator",
    "PolarsDataVersionCalculator",
    "DiffResult",
    "MetadataDiffResolver",
    "PolarsDiffResolver",
]
