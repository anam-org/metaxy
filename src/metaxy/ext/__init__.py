"""Integrations with third-party software."""
