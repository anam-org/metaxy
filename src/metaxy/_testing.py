import importlib
import sys
import tempfile
from functools import cached_property
from pathlib import Path
from typing import Any

from metaxy import (
    FeatureSpec,
)
from metaxy.config import MetaxyConfig
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.metadata_store.base import MetadataStore
from metaxy.models.feature import FeatureGraph


class TempFeatureModule:
    """Helper to create temporary Python modules with feature definitions.

    This allows features to be importable by historical graph reconstruction.
    The same import path (e.g., 'temp_features.Upstream') can be used across
    different feature versions by overwriting the module file.
    """

    def __init__(self, module_name: str = "temp_test_features"):
        self.temp_dir = tempfile.mkdtemp(prefix="metaxy_test_")
        self.module_name = module_name
        self.module_path = Path(self.temp_dir) / f"{module_name}.py"

        # Add to sys.path so module can be imported
        sys.path.insert(0, self.temp_dir)

    def write_features(self, feature_specs: dict[str, FeatureSpec]):
        """Write feature classes to the module file.

        Args:
            feature_specs: Dict mapping class names to FeatureSpec objects
        """
        code_lines = [
            "# Auto-generated test feature module",
            "from metaxy import Feature, FeatureSpec, FieldSpec, FieldKey, FeatureDep, FeatureKey, FieldDep, SpecialFieldDep",
            "from metaxy.models.feature import FeatureGraph",
            "",
            "# Use a dedicated graph for this temp module",
            "_graph = FeatureGraph()",
            "",
        ]

        for class_name, spec in feature_specs.items():
            # Generate the spec definition
            spec_dict = spec.model_dump(mode="python")
            spec_repr = self._generate_spec_repr(spec_dict)

            code_lines.extend(
                [
                    f"# Define {class_name} in the temp graph context",
                    "with _graph.use():",
                    f"    class {class_name}(",
                    "        Feature,",
                    f"        spec={spec_repr}",
                    "    ):",
                    "        pass",
                    "",
                ]
            )

        # Write the file
        self.module_path.write_text("\n".join(code_lines))

        # Reload module if it was already imported
        if self.module_name in sys.modules:
            importlib.reload(sys.modules[self.module_name])

    def _generate_spec_repr(self, spec_dict: dict[str, Any]) -> str:
        """Generate FeatureSpec constructor call from dict."""
        # This is a simple representation - could be made more robust
        parts = []

        # key
        key = spec_dict["key"]
        parts.append(f"key=FeatureKey({key!r})")

        # deps
        deps = spec_dict.get("deps")
        if deps is None:
            parts.append("deps=None")
        else:
            deps_repr = [f"FeatureDep(key=FeatureKey({d['key']!r}))" for d in deps]
            parts.append(f"deps=[{', '.join(deps_repr)}]")

        # fields
        fields = spec_dict.get("fields", [])
        if fields:
            field_reprs = []
            for c in fields:
                c_parts = [
                    f"key=FieldKey({c['key']!r})",
                    f"code_version={c['code_version']}",
                ]

                # Handle deps
                deps_val = c.get("deps")
                if deps_val == "__METAXY_ALL_DEP__":
                    c_parts.append("deps=SpecialFieldDep.ALL")
                elif isinstance(deps_val, list) and deps_val:
                    # Field deps (list of FieldDep)
                    cdeps: list[str] = []  # type: ignore[misc]
                    for cd in deps_val:
                        fields_val = cd.get("fields")
                        if fields_val == "__METAXY_ALL_DEP__":
                            cdeps.append(  # type: ignore[arg-type]
                                f"FieldDep(feature_key=FeatureKey({cd['feature_key']!r}), fields=SpecialFieldDep.ALL)"
                            )
                        else:
                            # Build list of FieldKey objects
                            field_keys = [f"FieldKey({k!r})" for k in fields_val]
                            cdeps.append(
                                f"FieldDep(feature_key=FeatureKey({cd['feature_key']!r}), fields=[{', '.join(field_keys)}])"
                            )
                    c_parts.append(f"deps=[{', '.join(cdeps)}]")

                field_reprs.append(f"FieldSpec({', '.join(c_parts)})")  # type: ignore[arg-type]

            parts.append(f"fields=[{', '.join(field_reprs)}]")

        return f"FeatureSpec({', '.join(parts)})"

    @property
    def graph(self) -> FeatureGraph:
        """Get the FeatureGraph from the temp module.

        Returns:
            The _graph instance from the imported module
        """
        # Import the module to get its _graph
        module = importlib.import_module(self.module_name)
        return module._graph

    def cleanup(self):
        """Remove temp directory and module from sys.path.

        NOTE: Don't call this until the test session is completely done,
        as historical graph loading may need to import from these modules.
        """
        if self.temp_dir in sys.path:
            sys.path.remove(self.temp_dir)

        # Remove from sys.modules
        if self.module_name in sys.modules:
            del sys.modules[self.module_name]

        # Delete temp directory
        import shutil

        shutil.rmtree(self.temp_dir, ignore_errors=True)


def assert_all_results_equal(results: dict[str, Any], snapshot=None) -> None:
    """Compare all results from different store type combinations.

    Ensures all variants produce identical results, then optionally snapshots all results.

    Args:
        results: Dict mapping store_type to result data
        snapshot: Optional syrupy snapshot fixture to record all results

    Raises:
        AssertionError: If any variants produce different results
    """
    if not results:
        return

    # Get all result values as a list
    all_results = list(results.items())
    reference_key, reference_result = all_results[0]

    # Compare each result to the reference
    for key, result in all_results[1:]:
        assert result == reference_result, (
            f"{key} produced different results than {reference_key}:\n"
            f"Expected: {reference_result}\n"
            f"Got: {result}"
        )

    # Snapshot ALL results if snapshot provided
    if snapshot is not None:
        assert results == snapshot


class HashAlgorithmCases:
    """Test cases for different hash algorithms."""

    def case_xxhash64(self) -> HashAlgorithm:
        """xxHash64 algorithm."""
        return HashAlgorithm.XXHASH64

    def case_xxhash32(self) -> HashAlgorithm:
        """xxHash32 algorithm."""
        return HashAlgorithm.XXHASH32

    def case_wyhash(self) -> HashAlgorithm:
        """WyHash algorithm."""
        return HashAlgorithm.WYHASH

    def case_sha256(self) -> HashAlgorithm:
        """SHA256 algorithm."""
        return HashAlgorithm.SHA256

    def case_md5(self) -> HashAlgorithm:
        """MD5 algorithm."""
        return HashAlgorithm.MD5


class TempMetaxyProject:
    """Helper for creating temporary Metaxy projects.

    Provides a context manager API for dynamically creating feature modules
    and running CLI commands with proper entrypoint configuration.

    Example:
        >>> project = TempMetaxyProject(tmp_path)
        >>>
        >>> def features():
        ...     from metaxy import Feature, FeatureSpec, FeatureKey, FieldSpec, FieldKey
        ...
        ...     class MyFeature(Feature, spec=FeatureSpec(
        ...         key=FeatureKey(["my_feature"]),
        ...         deps=None,
        ...         fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)]
        ...     )):
        ...         pass
        >>>
        >>> with project.with_features(features):
        ...     result = project.run_cli("graph", "push")
        ...     assert result.returncode == 0
    """

    def __init__(self, tmp_path: Path):
        """Initialize a temporary Metaxy project.

        Args:
            tmp_path: Temporary directory path (usually from pytest tmp_path fixture)
        """
        self.project_dir = tmp_path
        self.project_dir.mkdir(exist_ok=True)
        self._feature_modules: list[str] = []
        self._write_config()

    def _write_config(self):
        """Write basic metaxy.toml with DuckDB store configuration."""
        dev_db_path = self.project_dir / "metadata.duckdb"
        staging_db_path = self.project_dir / "metadata_staging.duckdb"
        config_content = f'''store = "dev"

[stores.dev]
type = "metaxy.metadata_store.duckdb.DuckDBMetadataStore"

[stores.dev.config]
database = "{dev_db_path}"

[stores.staging]
type = "metaxy.metadata_store.duckdb.DuckDBMetadataStore"

[stores.staging.config]
database = "{staging_db_path}"
'''
        (self.project_dir / "metaxy.toml").write_text(config_content)

    def with_features(self, features_func, module_name: str | None = None):
        """Context manager that sets up features for the duration of the block.

        Extracts source code from features_func (skipping the function definition line),
        writes it to a Python module file, and tracks it for METAXY_ENTRYPOINTS__N
        environment variable configuration.

        Args:
            features_func: Function containing feature class definitions.
                All imports must be inside the function body.
            module_name: Optional module name. If not provided, generates
                "features_N" based on number of existing modules.

        Yields:
            str: The module name that was created

        Example:
            >>> def my_features():
            ...     from metaxy import Feature, FeatureSpec, FeatureKey
            ...
            ...     class MyFeature(Feature, spec=...):
            ...         pass
            >>>
            >>> with project.with_features(my_features) as module:
            ...     print(module)  # "features_0"
            ...     result = project.run_cli("graph", "push")
        """
        import inspect
        import textwrap
        from contextlib import contextmanager

        @contextmanager
        def _context():
            # Generate module name if not provided
            nonlocal module_name
            if module_name is None:
                module_name = f"features_{len(self._feature_modules)}"

            # Extract source code from function
            source = inspect.getsource(features_func)

            # Remove function definition line and dedent
            lines = source.split("\n")
            # Find the first line that's not a decorator or function def
            body_start = 0
            for i, line in enumerate(lines):
                if line.strip().startswith("def ") and ":" in line:
                    body_start = i + 1
                    break

            body_lines = lines[body_start:]
            dedented = textwrap.dedent("\n".join(body_lines))

            # Write to file in project directory
            feature_file = self.project_dir / f"{module_name}.py"
            feature_file.write_text(dedented)

            # Track this module
            self._feature_modules.append(module_name)

            try:
                yield module_name
            finally:
                # Cleanup: remove from tracking (file stays for debugging)
                if module_name in self._feature_modules:
                    self._feature_modules.remove(module_name)

        return _context()

    def run_cli(
        self, *args, check: bool = True, env: dict[str, str] | None = None, **kwargs
    ):
        """Run CLI command with current feature modules loaded.

        Automatically sets METAXY_ENTRYPOINT_0, METAXY_ENTRYPOINT_1, etc.
        based on active with_features() context managers.

        Args:
            *args: CLI command arguments (e.g., "graph", "push")
            check: If True (default), raises CalledProcessError on non-zero exit
            env: Optional dict of additional environment variables
            **kwargs: Additional arguments to pass to subprocess.run()

        Returns:
            subprocess.CompletedProcess: Result of the CLI command

        Raises:
            subprocess.CalledProcessError: If check=True and command fails

        Example:
            >>> result = project.run_cli("graph", "history", "--limit", "5")
            >>> print(result.stdout)
        """
        import os
        import subprocess

        # Start with current environment
        cmd_env = os.environ.copy()

        # Add project directory to PYTHONPATH so modules can be imported
        pythonpath = str(self.project_dir)
        if "PYTHONPATH" in cmd_env:
            pythonpath = f"{pythonpath}{os.pathsep}{cmd_env['PYTHONPATH']}"
        cmd_env["PYTHONPATH"] = pythonpath

        # Set entrypoints for all tracked modules
        # Use METAXY_ENTRYPOINT_0, METAXY_ENTRYPOINT_1, etc. (single underscore for list indexing)
        for idx, module_name in enumerate(self._feature_modules):
            cmd_env[f"METAXY_ENTRYPOINT_{idx}"] = module_name

        # Apply additional env overrides
        if env:
            cmd_env.update(env)

        # Run CLI command
        result = subprocess.run(
            [sys.executable, "-m", "metaxy.cli.app", *args],
            cwd=str(self.project_dir),
            capture_output=True,
            text=True,
            env=cmd_env,
            check=check,
            **kwargs,
        )

        return result

    @property
    def entrypoints(self):
        return [f"METAXY_ENTRYPOINT_{idx}" for idx in range(len(self._feature_modules))]

    @property
    def graph(self) -> FeatureGraph:
        """Load features from the project's feature modules into a graph.

        Returns:
            FeatureGraph with all features from tracked modules loaded
        """
        import importlib
        import sys

        graph = FeatureGraph()

        # Ensure project dir is in sys.path
        project_dir_str = str(self.project_dir)
        was_in_path = project_dir_str in sys.path
        if not was_in_path:
            sys.path.insert(0, project_dir_str)

        try:
            with graph.use():
                # Import feature modules directly
                for module_name in self._feature_modules:
                    # Import or reload the module
                    if module_name in sys.modules:
                        importlib.reload(sys.modules[module_name])
                    else:
                        importlib.import_module(module_name)
        finally:
            # Clean up sys.path if we added it
            if not was_in_path and project_dir_str in sys.path:
                sys.path.remove(project_dir_str)

        return graph

    @cached_property
    def config(self) -> MetaxyConfig:
        return MetaxyConfig.load(self.project_dir / "metaxy.toml")

    @cached_property
    def stores(self) -> dict[str, MetadataStore]:
        return {k: self.config.get_store(k) for k in self.config.stores}
