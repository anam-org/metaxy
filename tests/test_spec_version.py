"""Tests for TestingFeatureSpec.feature_spec_version property."""

import hashlib
import json

from syrupy.assertion import SnapshotAssertion

from metaxy import (
    FeatureDep,
    FeatureKey,
    FieldKey,
    FieldSpec,
    TestingFeatureSpec,
)


def test_feature_spec_version_deterministic(snapshot: SnapshotAssertion) -> None:
    """Test that feature_spec_version is deterministic."""
    spec = TestingFeatureSpec(
        key=FeatureKey(["test", "feature"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )

    version1 = spec.feature_spec_version
    version2 = spec.feature_spec_version

    # Should be deterministic
    assert version1 == version2

    # Should be 64 characters (SHA256 hex digest)
    assert len(version1) == 64

    # Should be hex string
    assert all(c in "0123456789abcdef" for c in version1)

    # Snapshot the hash
    assert version1 == snapshot


def test_feature_spec_version_includes_all_properties(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that feature_spec_version includes all specification properties."""
    # Create a complex spec with all properties
    spec = TestingFeatureSpec(
        key=FeatureKey(["complex", "feature"]),
        deps=[
            FeatureDep(
                key=FeatureKey(["upstream", "one"]),
                columns=("col1", "col2"),
                rename={"col1": "renamed_col1"},
            ),
            FeatureDep(
                key=FeatureKey(["upstream", "two"]),
                columns=None,  # All columns
                rename=None,
            ),
        ],
        fields=[
            FieldSpec(key=FieldKey(["field1"]), code_version=1),
            FieldSpec(key=FieldKey(["field2"]), code_version=2),
        ],
    )

    version = spec.feature_spec_version

    # Verify it's a valid SHA256 hash
    assert len(version) == 64
    assert all(c in "0123456789abcdef" for c in version)

    # Snapshot for stability
    assert version == snapshot


def test_feature_spec_version_changes_with_any_property() -> None:
    """Test that feature_spec_version changes when any property changes."""
    base_spec = TestingFeatureSpec(
        key=FeatureKey(["base", "feature"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )
    base_version = base_spec.feature_spec_version

    # Change key
    spec_key_changed = TestingFeatureSpec(
        key=FeatureKey(["changed", "feature"]),  # Changed!
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )
    assert spec_key_changed.feature_spec_version != base_version

    # Add deps
    spec_deps_added = TestingFeatureSpec(
        key=FeatureKey(["base", "feature"]),
        deps=[FeatureDep(key=FeatureKey(["upstream"]))],  # Added!
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )
    assert spec_deps_added.feature_spec_version != base_version

    # Change field code_version
    spec_field_changed = TestingFeatureSpec(
        key=FeatureKey(["base", "feature"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=2),  # Changed!
        ],
    )
    assert spec_field_changed.feature_spec_version != base_version

    # Add field
    spec_field_added = TestingFeatureSpec(
        key=FeatureKey(["base", "feature"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
            FieldSpec(key=FieldKey(["new_field"]), code_version=1),  # Added!
        ],
    )
    assert spec_field_added.feature_spec_version != base_version


def test_feature_spec_version_consistent_ordering() -> None:
    """Test that feature_spec_version is consistent regardless of field order."""
    # Create specs with fields in different order
    spec1 = TestingFeatureSpec(
        key=FeatureKey(["test", "ordering"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["field_a"]), code_version=1),
            FieldSpec(key=FieldKey(["field_b"]), code_version=2),
        ],
    )

    spec2 = TestingFeatureSpec(
        key=FeatureKey(["test", "ordering"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["field_b"]), code_version=2),  # Reordered
            FieldSpec(key=FieldKey(["field_a"]), code_version=1),
        ],
    )

    # The order of fields in the list matters for the spec
    # So these should have different hashes
    assert spec1.feature_spec_version != spec2.feature_spec_version

    # But creating the same spec multiple times should be deterministic
    spec3 = TestingFeatureSpec(
        key=FeatureKey(["test", "ordering"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["field_a"]), code_version=1),
            FieldSpec(key=FieldKey(["field_b"]), code_version=2),
        ],
    )
    assert spec1.feature_spec_version == spec3.feature_spec_version


def test_feature_spec_version_manual_verification() -> None:
    """Manually verify the feature_spec_version computation."""
    spec = TestingFeatureSpec(
        key=FeatureKey(["manual", "test"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )

    # Get the spec version
    feature_spec_version = spec.feature_spec_version

    # Manually compute what it should be
    spec_dict = spec.model_dump(mode="json")
    spec_json = json.dumps(spec_dict, sort_keys=True)
    expected_hash = hashlib.sha256(spec_json.encode("utf-8")).hexdigest()

    # They should match
    assert feature_spec_version == expected_hash


def test_feature_spec_version_with_column_selection_and_rename(
    snapshot: SnapshotAssertion,
) -> None:
    """Test feature_spec_version with column selection and renaming in deps."""
    spec = TestingFeatureSpec(
        key=FeatureKey(["test", "columns"]),
        deps=[
            FeatureDep(
                key=FeatureKey(["upstream"]),
                columns=("col1", "col2"),
                rename={"col1": "new_col1", "col2": "new_col2"},
            )
        ],
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )

    version = spec.feature_spec_version

    # Should be valid SHA256
    assert len(version) == 64
    assert all(c in "0123456789abcdef" for c in version)

    # Snapshot for stability
    assert version == snapshot

    # Changing column selection should change the hash
    spec_different_columns = TestingFeatureSpec(
        key=FeatureKey(["test", "columns"]),
        deps=[
            FeatureDep(
                key=FeatureKey(["upstream"]),
                columns=("col1", "col3"),  # Different columns!
                rename={"col1": "new_col1", "col3": "new_col3"},
            )
        ],
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )
    assert spec_different_columns.feature_spec_version != version

    # Changing rename mapping should change the hash
    spec_different_rename = TestingFeatureSpec(
        key=FeatureKey(["test", "columns"]),
        deps=[
            FeatureDep(
                key=FeatureKey(["upstream"]),
                columns=("col1", "col2"),
                rename={"col1": "different_name"},  # Different rename!
            )
        ],
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    )
    assert spec_different_rename.feature_spec_version != version


def test_feature_feature_spec_version_classmethod() -> None:
    """Test that Feature.feature_spec_version() classmethod works correctly."""
    from metaxy import Feature, FeatureGraph

    graph = FeatureGraph()

    with graph.use():

        class TestFeature(
            Feature,
            spec=TestingFeatureSpec(
                key=FeatureKey(["test", "classmethod"]),
                deps=None,
                fields=[
                    FieldSpec(key=FieldKey(["default"]), code_version=1),
                ],
            ),
        ):
            pass

        # Get feature_spec_version via classmethod
        classmethod_version = TestFeature.feature_spec_version()

        # Get feature_spec_version directly from spec
        direct_version = TestFeature.spec().feature_spec_version

        # They should be identical
        assert classmethod_version == direct_version

        # Should be valid SHA256
        assert len(classmethod_version) == 64
        assert all(c in "0123456789abcdef" for c in classmethod_version)


def test_feature_spec_version_stored_in_snapshot(snapshot: SnapshotAssertion) -> None:
    """Test that feature_spec_version is included in graph snapshots."""
    from metaxy import Feature, FeatureGraph

    graph = FeatureGraph()

    with graph.use():

        class SnapshotFeature(
            Feature,
            spec=TestingFeatureSpec(
                key=FeatureKey(["snapshot", "test"]),
                deps=None,
                fields=[
                    FieldSpec(key=FieldKey(["data"]), code_version=1),
                ],
            ),
        ):
            pass

        # Get snapshot dict
        snapshot_dict = graph.to_snapshot()

        # Should have the feature
        feature_key_str = "snapshot/test"
        assert feature_key_str in snapshot_dict

        # Should contain feature_spec_version
        feature_data = snapshot_dict[feature_key_str]
        assert "feature_spec_version" in feature_data

        # feature_spec_version should match the Feature's feature_spec_version
        assert (
            feature_data["feature_spec_version"]
            == SnapshotFeature.spec().feature_spec_version
        )

        # feature_spec_version should be different from feature_version
        assert feature_data["feature_spec_version"] != feature_data["feature_version"]

        # Snapshot the entire feature data for stability
        assert feature_data == snapshot


def test_feature_spec_version_recorded_in_metadata_store(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that feature_spec_version is recorded when pushing to metadata store."""

    from metaxy import Feature, FeatureGraph
    from metaxy.metadata_store import InMemoryMetadataStore

    graph = FeatureGraph()

    with graph.use():

        class RecordedFeature(
            Feature,
            spec=TestingFeatureSpec(
                key=FeatureKey(["recorded", "feature"]),
                deps=None,
                fields=[
                    FieldSpec(key=FieldKey(["default"]), code_version=1),
                ],
            ),
        ):
            pass

        store = InMemoryMetadataStore()

        with store:
            # Record the feature graph snapshot
            result = store.record_feature_graph_snapshot()

            is_existing = result.already_recorded

            # Should be a new snapshot
            assert not is_existing

            # Read the feature_versions system table
            features_df = store.read_features(current=True)

            # Should have one feature
            assert len(features_df) == 1

            # Check that feature_spec_version column exists and has value
            assert "feature_spec_version" in features_df.columns

            # Get row as dict
            feature_row = features_df.to_dicts()[0]
            assert feature_row["feature_spec_version"] is not None
            assert len(feature_row["feature_spec_version"]) == 64

            # feature_spec_version should match the Feature's feature_spec_version
            assert (
                feature_row["feature_spec_version"]
                == RecordedFeature.spec().feature_spec_version
            )

            # feature_spec_version should be different from feature_version
            assert feature_row["feature_spec_version"] != feature_row["feature_version"]

            # Snapshot for stability
            assert {
                "feature_key": feature_row["feature_key"],
                "feature_version": feature_row["feature_version"],
                "feature_spec_version": feature_row["feature_spec_version"],
                "snapshot_version": feature_row["snapshot_version"],
            } == snapshot


def test_feature_spec_version_idempotent_snapshot_recording() -> None:
    """Test that recording the same snapshot twice preserves feature_spec_version."""
    from metaxy import Feature, FeatureGraph
    from metaxy.metadata_store import InMemoryMetadataStore

    graph = FeatureGraph()

    with graph.use():

        class IdempotentFeature(
            Feature,
            spec=TestingFeatureSpec(
                key=FeatureKey(["idempotent", "test"]),
                deps=None,
                fields=[
                    FieldSpec(key=FieldKey(["default"]), code_version=1),
                ],
            ),
        ):
            pass

        store = InMemoryMetadataStore()

        with store:
            # First push
            result = store.record_feature_graph_snapshot()

            snapshot_v1 = result.snapshot_version

            is_existing_1 = result.already_recorded
            assert not is_existing_1

            features_df_1 = store.read_features(current=True)
            feature_spec_version_1 = features_df_1.to_dicts()[0]["feature_spec_version"]

            # Second push (identical graph)
            result = store.record_feature_graph_snapshot()

            snapshot_v2 = result.snapshot_version

            is_existing_2 = result.already_recorded
            assert is_existing_2  # Should detect existing snapshot
            assert snapshot_v1 == snapshot_v2  # Same snapshot version

            features_df_2 = store.read_features(current=True)
            feature_spec_version_2 = features_df_2.to_dicts()[0]["feature_spec_version"]

            # feature_spec_version should be identical
            assert feature_spec_version_1 == feature_spec_version_2


def test_feature_spec_version_different_from_feature_version_always() -> None:
    """Test that feature_spec_version is always different from feature_version.

    These two hashes serve different purposes and use different computation methods:
    - feature_spec_version: Direct JSON serialization of TestingFeatureSpec (all properties)
    - feature_version: Graph-based computation including dependency chains
    """
    from metaxy import Feature, FeatureGraph

    graph = FeatureGraph()

    with graph.use():
        # Test with root feature (no deps)
        class RootFeature(
            Feature,
            spec=TestingFeatureSpec(
                key=FeatureKey(["root"]),
                deps=None,
                fields=[
                    FieldSpec(key=FieldKey(["default"]), code_version=1),
                ],
            ),
        ):
            pass

        # Test with downstream feature (with deps)
        class DownstreamFeature(
            Feature,
            spec=TestingFeatureSpec(
                key=FeatureKey(["downstream"]),
                deps=[FeatureDep(key=FeatureKey(["root"]))],
                fields=[
                    FieldSpec(key=FieldKey(["default"]), code_version=1),
                ],
            ),
        ):
            pass

        # Both should have different feature_spec_version vs feature_version
        assert RootFeature.feature_spec_version() != RootFeature.feature_version()
        assert (
            DownstreamFeature.feature_spec_version()
            != DownstreamFeature.feature_version()
        )

        # Both versions should be valid SHA256 hashes
        for feature_cls in [RootFeature, DownstreamFeature]:
            spec_v = feature_cls.feature_spec_version()
            feature_v = feature_cls.feature_version()

            assert len(spec_v) == 64
            assert len(feature_v) == 64
            assert all(c in "0123456789abcdef" for c in spec_v)
            assert all(c in "0123456789abcdef" for c in feature_v)


def test_feature_spec_version_with_multiple_complex_deps(
    snapshot: SnapshotAssertion,
) -> None:
    """Test feature_spec_version with multiple complex dependencies."""
    spec = TestingFeatureSpec(
        key=FeatureKey(["complex", "multi", "dep"]),
        deps=[
            FeatureDep(
                key=FeatureKey(["upstream", "one"]),
                columns=("a", "b", "c"),
                rename={"a": "upstream1_a", "b": "upstream1_b"},
            ),
            FeatureDep(
                key=FeatureKey(["upstream", "two"]),
                columns=(),  # Empty tuple = only system columns
                rename=None,
            ),
            FeatureDep(
                key=FeatureKey(["upstream", "three"]),
                columns=None,  # None = all columns
                rename={"col": "renamed_col"},
            ),
        ],
        fields=[
            FieldSpec(key=FieldKey(["field1"]), code_version=1),
            FieldSpec(key=FieldKey(["field2"]), code_version=2),
            FieldSpec(key=FieldKey(["field3"]), code_version=3),
        ],
    )

    version = spec.feature_spec_version

    # Should be valid SHA256
    assert len(version) == 64
    assert all(c in "0123456789abcdef" for c in version)

    # Snapshot the hash
    assert version == snapshot

    # Verify the spec is deterministic
    spec_copy = TestingFeatureSpec(
        key=FeatureKey(["complex", "multi", "dep"]),
        deps=[
            FeatureDep(
                key=FeatureKey(["upstream", "one"]),
                columns=("a", "b", "c"),
                rename={"a": "upstream1_a", "b": "upstream1_b"},
            ),
            FeatureDep(
                key=FeatureKey(["upstream", "two"]),
                columns=(),
                rename=None,
            ),
            FeatureDep(
                key=FeatureKey(["upstream", "three"]),
                columns=None,
                rename={"col": "renamed_col"},
            ),
        ],
        fields=[
            FieldSpec(key=FieldKey(["field1"]), code_version=1),
            FieldSpec(key=FieldKey(["field2"]), code_version=2),
            FieldSpec(key=FieldKey(["field3"]), code_version=3),
        ],
    )

    assert spec_copy.feature_spec_version == version
