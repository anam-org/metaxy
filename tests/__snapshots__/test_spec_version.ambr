# serializer version: 1
# name: test_feature_spec_version_deterministic
  '90179e7bd99745bd2494f695fa18c15937353a3788c9b5c4650a81a071c1f591'
# ---
# name: test_feature_spec_version_includes_all_properties
  'e7c76b7981649ad1e0d9c8f373043eb42ef75feb7f1a5e222236afb632f12e97'
# ---
# name: test_feature_spec_version_recorded_in_metadata_store
  dict({
    'feature_key': 'recorded/feature',
    'feature_spec_version': 'd64c6d4dcd782fda8988b6f789367a2c10aed2732e350eb3e2a285051c73d493',
    'feature_version': '0c7b2d83252fbf2f689bec7d37c4f7ffb103755d11eafabc1e000a372f415d83',
    'snapshot_version': '14d4294da40f4ab27ca85eb739fffd70e92c92522dbcba995cfa2aa343988bf3',
  })
# ---
# name: test_feature_spec_version_stored_in_snapshot
  dict({
    'feature_class_path': 'tests.test_spec_version.SnapshotFeature',
    'feature_spec': dict({
      'deps': None,
      'fields': list([
        dict({
          'code_version': 1,
          'deps': '__METAXY_ALL_DEP__',
          'key': list([
            'data',
          ]),
        }),
      ]),
      'id_columns': list([
        'sample_uid',
      ]),
      'key': list([
        'snapshot',
        'test',
      ]),
      'metadata': dict({
      }),
    }),
    'feature_spec_version': '6dc07209be338e7e86bb47dc3f376f3e7447efe35510cac9119ee0878c8c20a2',
    'feature_tracking_version': 'af52f7692de124d551c0753a169b2a0b4a57003818ed08474c6d9c17480b7e20',
    'feature_version': '7cfde77960e3cf10327e9cb97f311418b101fc5ee2c146922a997306e318edaf',
    'project': 'test',
  })
# ---
# name: test_feature_spec_version_with_column_selection_and_rename
  'cf22f9791d5b83dc084c493e52914a21c6fe27d261e78ec03fe22e033f750b4a'
# ---
# name: test_feature_spec_version_with_multiple_complex_deps
  '20703ec5548c00dbc3f0d634a6791f284d6b1cec4701d014fe5ccaec4deef74b'
# ---
